src/app/rosetta/lib/account.ml

```diff
@@ -29,14 +29,36 @@ module Balance_info = struct
            [@@deriving yojson]
 end
 
+
 module Sql = struct
   module Balance_from_last_relevant_command = struct
+    let max_txns =
+      Int.pow 2 Genesis_constants.Constraint_constants.compiled.transaction_capacity_log_2
 
     let query_pending =
       Caqti_request.find_opt
         Caqti_type.(tup2 string int64)
-        Caqti_type.(tup3 int64 int64 (option int64))
-        {sql| WITH RECURSIVE pending_chain AS (
+        Caqti_type.(tup4 int64 int64 int64 (option int64))
+        (sprintf
+        {sql|
+SELECT DISTINCT
+  combo.pk_id, -- this is only used as a slug to combine the rows but ignored in the OCaml
+  MAX(combo.block_global_slot_since_genesis) AS block_global_slot_since_genesis,
+  MAX(combo.balance) AS balance,
+  MAX(combo.nonce) AS nonce
+FROM (
+  /* There are two large recursive subqueries here. One for balance, and the
+   * other for nonce.
+   *
+   * These are separate subqueries because there is a quirk where a transaction
+   * can be received and sent in the same block and the latest nonce is not
+   * quite stored properly. To work around that, we are going to query _almost_
+   * the same thing but take the MAX nonce among the most recent 255 entries
```

Comment:
probably shouldn't mention the limit, even in a comment, since it may later change